runtime/doc/lsp.txt

```diff
@@ -0,0 +1,265 @@
+*lsp.txt* The Language Server Protocol
+
+        NVIM REFERENCE MANUAL
+
+
+Neovim Language Server Protocol (LSP) API
+
+Neovim exposes a powerful API that conforms to Microsoft's published Language
+Server Protocol specification. The documentation can be found here:
+
+    https://github.com/Microsoft/language-server-protocol
+
+================================================================================
+                                                                  *lsp-request*
+Requests
+
+To make a request of the language server client running in Nvim, the following
+API is defined:
+>
+    :call lsp#request(method, arguments, callback, filetype)
+    :call lsp#request_async(method, arguments, callback, filetype)
+<
+
+                                                                *lsp#request()*
+lsp#request({method} [, {arguments} [, {callback} [, {filetype}]]])
+    Creates and sends a request to the server for [filetype]. Will fill in any
+    missing required arguments as defined by the language server protocol. Will
+    call the default callback for the request (if it exists), unless an
+    [callback] is specified, in which case that will be called upon completion
+    of the request.
+
+    Returns the request's response data.
+
+        Parameters:~
+            {method}: The name of the request, as specified by the
+            language server protocol.
+                Example: 'textDocument/hover'
+
+            [arguments]: The arguments that should be passed to the language
+            server. If any required keys are missing, Nvim will fill those in
+            with reasonable default values.
+                Example (with 'textDocument/hover'):
+                    Passed in:
+                        `{}` or |v:null|
+
+                    Sent:
+>
+                        {
+                            'textDocument': '/home/user/test/example.lua',
+                            'position': { 'line': 13, 'character': 17 }
+                        }
+<
+                Example (with 'textDocument/rename'):
+                    Passed in:
+>
+                        {
+                            'newName': 'exampleRename'
+                        }
+<
+                    Sent:
+>
+                        {
+                            'textDocument': '/home/user/test/rename.lua',
+                            'position': { 'line': 16, 'character': 8 }
+                            'newName': 'exampleRename'
+                        }
+<
+
+            [callback]: An optional callback that will be executed
+            once the response has been received. If this is passed and not
+            |v:null|, then the |lsp-default-callback| will not be executed for
```

Comment:
IMHO this double usage of this function seems a bit confusion of purpose/abstraction. I would expect a low-level function like this to _only_ serialize a request and then either (sync) return the result or (async) invoke a specified callback. Opionated behavior is better defined in a seprarate module (but shipping some simple behaviors with runtime probably makes sense), and say invoked by commands like `:Lsp rename` or `LspRename`. Because opionated behavior might also be needed to prepare the request itself, like prompting the user for the new name with `rename`. Also consumer plugins can define functions/commands with their own namespaces, rather than fighting over the default callback or pre/post autocmds.